File: Functions/Stop_Criterion_Network_3D.cpp
#include <iostream>
#include <cmath>
#include</home/smm5969/Desktop/cpp/Eigen/Eigen/Dense>
#include</home/smm5969/Desktop/cpp/Eigen/Eigen/Sparse>
#include <tuple>
#include <vector>
using namespace std;
using namespace Eigen;
typedef Matrix <int, Dynamic, Dynamic> MatrixXi; // define matrix for integers.
int Stop_Criterion_Network_3D(MatrixXd P,MatrixXd Q_pre,MatrixXd Q, MatrixXd R,MatrixXd Fluxes,MatrixXd d,MatrixXi DOF,MatrixXd D_DOF,double u0,double tol_D,double tol_R,int choice )
{   int Stop_Key;
	if (choice==1) // Residue criterion
	{
		// double Ratio_R;
        // Ratio_R = R.norm()/R0.norm();
        // cout << "Norm_R:  " << R.norm() << endl;
        // cout << "Ratio_R:  " << Ratio_R << endl;
        // if (Ratio_R < tol_R)
        // {
            	// Stop_Key=1;
		// }
        // else
        // {
           // Stop_Key=0;	
	    // }

	}
	if (choice==2) // Displacement criterion
	{   
		// int num_disp,num_rot;
		// num_disp = 3*P.cols();num_rot = P.cols();
		// RowVectorXd d0_disp(num_disp),d0_rot(num_rot),d_disp(num_disp),d_rot(num_rot);
		// for (int j=0;j<Q0.cols();j++)
        // {   
        	// d0_disp(3*j)=Q0(0,j)-Q(0,j);d0_disp(3*j+1)=Q0(1,j)-Q(1,j);d0_disp(3*j+2)=Q0(2,j)-Q(2,j);
            // d0_rot(j)=Q0(3,j)-Q(3,j); 
            // d_disp(3*j)=d(DOF(j,0)-1,0);d_disp(3*j+1)=d(DOF(j,1)-1,0);d_disp(3*j+2)=d(DOF(j,2)-1,0);
            // d_rot(j)=d(DOF(j,3)-1,0); 
		// }
		// double Ratio_D_disp,Ratio_D_rot;
        // Ratio_D_disp = d_disp.norm()/d0_disp.norm();
        // Ratio_D_rot = d_rot.norm()/d0_rot.norm();
        // //cout << "d0_disp:  " << d0_disp << endl;
        // //cout << "d0_rot:  " << d0_rot << endl;
        // cout << "Ratio_D_disp:  " << Ratio_D_disp << endl;
        // cout << "Ratio_D_rot:  " << Ratio_D_rot << endl;
        // if (Ratio_D_disp < tol_D)
        // {
            	// Stop_Key=1;
		// }
        // else
        // {
           // Stop_Key=0;	
	    // } 

	} 
	if (choice==3) // Both displacement and residue criterion
	{
        MatrixXd flux_disp,flux_rot,r_disp,r_rot,d_disp,d_rot,Q_diff_disp,Q_diff_rot;
        Q_diff_disp.resize(3,3*P.cols()); Q_diff_rot.resize(P.cols(),1);
        double avg_flux_disp,avg_flux_rot,ratio_R_disp,ratio_R_rot,ratio_U_disp,ratio_U_rot,Ratio_D,Ratio_R;
        double key_disp,key_rot,temp_disp,temp_rot;
        
        flux_disp.resize(3*P.cols(),1);flux_rot.resize(P.cols(),1);
        r_disp.resize(3*P.cols(),1);r_rot.resize(P.cols(),1);
        d_disp.resize(3*P.cols(),1);d_rot.resize(P.cols(),1);
                
        // Comparing residual to the fluxes for residual criterion
        // Checking how far we have come from assumed solution
        //Separate different types of dofs (displacement and rotation)
        key_disp=0; key_rot=0;temp_disp=0; temp_rot=0;
        
        for (int i=0;i<P.cols();i++)
        {
                for (int j=0;j<3;j++)
            {
                if (D_DOF(i,j)!=u0)
                {
                    // temp_disp+=abs(Fluxes(DOF(i,j)-1,0));key_disp+=1;
                    temp_disp+=pow(Fluxes(DOF(i,j)-1,0),2);
                }	
                
            }
            if (D_DOF(i,3)!=u0)
            {
                    // temp_rot+=abs(Fluxes(DOF(i,3)-1,0));key_rot+=1;
                    temp_rot+=pow(Fluxes(DOF(i,3)-1,0),2);
            } 
            flux_disp.block(3*i,0,3,1)=Fluxes.block(4*i,0,3,1);flux_rot(i,0)=Fluxes(4*i+3,0);
            r_disp.block(3*i,0,3,1)=R.block(4*i,0,3,1);r_rot(i,0)=R(4*i+3,0);
            d_disp.block(3*i,0,3,1)=d.block(4*i,0,3,1);d_rot(i,0)=d(4*i+3,0);
            Q_diff_disp.block(0,i,3,1)=Q_pre.block(0,i,3,1)-Q.block(0,i,3,1);
            Q_diff_rot(i,0)=Q_pre(3,i)-Q(3,i);
        }
        flux_disp=flux_disp.cwiseAbs();flux_rot=flux_rot.cwiseAbs();
        r_disp=r_disp.cwiseAbs();r_rot=r_rot.cwiseAbs();
        d_disp=d_disp.cwiseAbs();d_rot=d_rot.cwiseAbs();
        avg_flux_disp=flux_disp.norm();avg_flux_rot=flux_rot.norm();
        // cout << "avg_flux_disp" << avg_flux_disp << endl;
        // avg_flux_disp=temp_disp/key_disp;avg_flux_rot=temp_rot/key_rot;
        // avg_flux_disp=pow(temp_disp,0.5);avg_flux_rot=pow(temp_rot,0.5);
        // cout << "avg_flux_disp" << avg_flux_disp << endl;
        // cout << "r_disp.maxCoeff()" << r_disp.maxCoeff() << endl;
        ratio_R_disp=r_disp.maxCoeff()/avg_flux_disp;
        ratio_R_rot=r_rot.maxCoeff()/avg_flux_rot;
        
        Q_diff_disp=Q_diff_disp.cwiseAbs();Q_diff_rot=Q_diff_rot.cwiseAbs();
        d_disp=d_disp.cwiseAbs();d_rot=d_rot.cwiseAbs();
        ratio_U_disp=d_disp.maxCoeff()/Q_diff_disp.maxCoeff();
        ratio_U_rot=d_rot.maxCoeff()/Q_diff_rot.maxCoeff();

        // only considering displacement dof no rotation
        Ratio_D=ratio_U_disp;Ratio_R=ratio_R_disp;
        
        // cout << " ratio_R_disp " << ratio_R_disp << endl;
        // cout << " ratio_R_rot " << ratio_R_rot << endl;
        // cout << " ratio_U_disp " << ratio_U_disp << endl;
        // cout << " ratio_U_rot " << ratio_U_rot << endl;

        
        if (Ratio_D < tol_D && Ratio_R < tol_R)
        {
            Stop_Key=1;
            cout << "Ratio_D " << Ratio_D << endl;
            cout << "Ratio_R " << Ratio_R << endl;
        }
        else
        {
            Stop_Key=0;
            cout << "Ratio_D " << Ratio_D << endl;
            cout << "Ratio_R " << Ratio_R << endl;
        }
	}
	return Stop_Key;
}

File: Matlab/Formulation/German Formulation/With Torsion/Functions_RN Parallel/3D_New/Stop_Criterion_Network_3D.m
function [Stop_Key] = Stop_Criterion_Network_3D(P,Q0,Q,d,R,Patch,Patch_Bs,A0,A0_Bs,Mat,Mat_Arcs,Mat_Bs,DOF,DOF_Bs,tol_D,tol_R,choice)
%% Stopping Criterion: Based on residue or displacement or both
%% Input
% P :
% Q1 :
% Q2:
% R :
% tol_D :
% tol_R :
%% Output
%% Code
% Area=Mat_Bs(1);E_bs=Mat_Bs(2);Ip=Mat_Bs(6);r=10^-3;
%% 
ncp=size(DOF,1); % no of control points
switch choice
%% Residue only
    case 1
        Ratio_R=Norm_R(length(Norm_R))/Norm_R(1);
        if Ratio_R<tol_R
            Stop_Key=1;
        else
            Stop_Key=0;
        end

%% Displacement Only
    case 2
        d0_disp=zeros((3/4)*size(R,1),1);d0_rot=zeros((1/4)*size(R,1),1);
        d_disp=zeros((3/4)*size(R,1),1);d_rot=zeros((3/4)*size(R,1),1);
        for j=1:size(Q0,2)
            d0_disp(3*(j-1)+1,1)=Q0(1,j)-Q(1,j);d0_disp(3*(j-1)+2,1)=Q0(2,j)-Q(2,j);d0_disp(3*(j-1)+3,1)=Q0(3,j)-Q(3,j);
            d0_rot(j,1)=Q0(4,j)-Q(4,j); 
            d_disp(3*(j-1)+1,1)=d(DOF(j,1),1);d_disp(3*(j-1)+2,1)=d(DOF(j,2),1);d_disp(3*(j-1)+3,1)=d(DOF(j,3),1);
            d_rot(j,1)=d(DOF(j,4),1);    
        end
        Ratio_D_disp = norm(d_disp)/norm(d0_disp);
%         Ratio_D_disp = max(abs(d_disp))/max(abs(d0_disp))
        if Ratio_D_disp < tol_D
            Stop_Key=1;
        else
            Stop_Key=0;
        end

%% Both displacement and residue
    case 3
        % Comparing residual to the fluxes for residual criterion
        % Checking how far we have come from assumed solution 
            [ Flux ] = Network_R_New(P,Q,Patch,Patch_Bs,A0,A0_Bs,Mat,Mat_Arcs,Mat_Bs,DOF,DOF_Bs);Flux=full(Flux);
            flux_disp=zeros(3*ncp,1);flux_rot=zeros(ncp,1);r_disp=zeros(ncp*3,1);r_rot=zeros(ncp,1);
            d_disp=zeros(3*ncp,1);d_rot=zeros(ncp,1);
            for k=1:ncp
                flux_disp(3*(k-1)+1:3*(k-1)+3,1)=Flux(DOF(k,1:3),1);flux_rot(k,1)=Flux(DOF(k,4),1);
                r_disp(3*(k-1)+1:3*(k-1)+3,1)=R(DOF(k,1:3),1);r_rot(k,1)=R(DOF(k,4),1);
                d_disp(3*(k-1)+1:3*(k-1)+3,1)=d(DOF(k,1:3),1);d_rot(k,1)=d(DOF(k,4),1);
            end
            avg_flux_disp=sum(abs(flux_disp))/length(flux_disp);avg_flux_rot=sum(abs(flux_rot))/length(flux_rot);
            ratio_R_disp=max(abs(r_disp))/avg_flux_disp
            ratio_R_rot=max(abs(r_rot))/avg_flux_rot
            ratio_U_disp=max(abs(d_disp))/max(max(abs(Q0(:,1:3)-Q(:,1:3))));
            ratio_U_rot=max(abs(d_rot))/max(max(abs(Q0(:,4)-Q(:,4))));
            
            % Ratio_D=max(ratio_disp,ratio_rot); Ratio_R=max(ratio_disp,ratio_rot); %both displacement and torsion
            
            Ratio_R=ratio_R_disp % only for displacement field
            Ratio_D=ratio_U_disp % only for displacement field
        if Ratio_D < tol_D && Ratio_R < tol_R
            Stop_Key=1;
        else
            Stop_Key=0;
        end  
end
end















